original/man/man3/vfdt.3
.TH "vfdt" 3 "28 Jul 2003" "VFML" \" -*- nroff -*-
.ad l
.nh
.SH NAME
vfdt \- 
.SH "Detailed Description"
.PP 
Learns a decision tree from a high-speed data stream or very large data set. 

vfdt is described in \fCthis paper\fP. (This version of VFDT has many extensions since that paper was written, including the ability to learn from domains with continuous attributes, we hope to have a more up-to-date paper to cite here soon).
.PP
VFDT learns a decision tree as follows. It starts with a single leaf and starts collecting training examples from a data stream (with the -stdin argument) or from the file stem.data. When it has enough data to know, with high confidence (see the -sc parameter below) that it knows which attribute is the best to partition the data with, it turns the leaf into an internal node splitting on the selected attribute and starts learning at the new leaves recursively. VFDT is an incremental online algoriithm in that it has a model available at any time during its run and refines the model over time as it is presented with additional training data (see the \fBvfdt-engine::h\fP interface, for an API to the learning engine if you want to incrementially learn decision trees in your own programs). VFDT will cache training examples in RAM if it has enough memory available (see the -growMegs parameter below) or it will just use them to update the statistics in the leaf where they belong and then free them. VFDT will also disable growing at unpromising leaves (and free the associated sufficient statistics) to save additional RAM when needed.
.PP
VFDT will be most effective when learning from very large data sets, in the millions or billions, where there is plenty of data for it to make good statistical decisions. For smaller data sets you may consider the -batch parameter which makes it load all data into RAM and learn as a traditional decision tree learner. If you suspect that there is concept drift in the data set you may like to use the \fBcvfdt\fP tool instead.
.PP
For a more detailed example of vfdt in action see the \fCmining data streams\fP walkthrough.
.PP
vfdt takes input and does output in \fCc4.5 format\fP. It expects to find the files \fC<stem>.names\fP and \fC<stem>.data\fP.
.PP
.SS "Arguments"
.PP
.IP "\(bu" 2
-f 'filestem'
.IP "  \(bu" 4
Set the name of the dataset (default DF)
.PP

.IP "\(bu" 2
-source 'dir'
.IP "  \(bu" 4
Set the source data directory (default '.')
.PP

.IP "\(bu" 2
-u
.IP "  \(bu" 4
Test the learner's accuracy on data in 'stem'.test
.PP

.IP "\(bu" 2
-sc 'prob'
.IP "  \(bu" 4
Allowed chance of error in each decision (default 0.0000001 that's .00001 percent)
.PP

.IP "\(bu" 2
-tc 'tie error' Call a tie when hoeffding e less than this. (default 0.05)
.IP "\(bu" 2
-gini
.IP "  \(bu" 4
Use gini gain instead of information gain (default information gain) (DOESN'T WORK FOR CONTINUOUS ATTRIBUTES)
.PP

.IP "\(bu" 2
-rescans '#'
.IP "  \(bu" 4
Naievely consider each example '#' times with no concern for using it once per level of the induced tree (default 1)
.PP

.IP "\(bu" 2
-chunk 'count'
.IP "  \(bu" 4
Wait until 'count' examples accumulate at a leaf before testing for a split (default 300)
.PP

.IP "\(bu" 2
-growMegs 'count'
.IP "  \(bu" 4
Limit dynamic memory allocation to 'count' megabytes (default 1000)
.PP

.IP "\(bu" 2
-stdin
.IP "  \(bu" 4
Reads training examples from stdin instead of from 'stem'.data (default off) NOTE this disables the rescans switch
.PP

.IP "\(bu" 2
-schedule 'mult'
.IP "  \(bu" 4
Run tests after 10k examples. Then sets the next test point to the current num examples * mult.
.PP

.IP "\(bu" 2
-outputTree
.IP "  \(bu" 4
Output the binary tree at each test point.
.PP

.IP "\(bu" 2
-noCacheTestSet
.IP "  \(bu" 4
By default vfdt reads test set from disk once and keeps in memory. Use this option to conserve some memory.
.PP

.IP "\(bu" 2
-prune
.IP "  \(bu" 4
Do reduced error pruning on the resulting tree, reserving 5% of data up to 2 million samples for pruning (default do not).
.PP

.IP "\(bu" 2
-noRestartLeaves
.IP "  \(bu" 4
Do not attempt to restart leaves that were deactivated for memory reasons (default do).
.PP

.IP "\(bu" 2
-noCacheTrainingExamples
.IP "  \(bu" 4
Do not use extra RAM to cache training examples (default do).
.PP

.IP "\(bu" 2
-cachePruneExamples
.IP "  \(bu" 4
Use extra RAM to cache pruning examples only makes sense if you use -prune (default store prune data in 'stem'.prunedata)
.PP

.IP "\(bu" 2
-batch
.IP "  \(bu" 4
Run in batch mode, read all examples into RAM, don't do hoeffding bounds (default off).
.PP

.IP "\(bu" 2
-laplace 'count'
.IP "  \(bu" 4
When starting a new node use the laplace correction with the parent's class and 'count' examples (defult 5)
.PP

.IP "\(bu" 2
-prePruneTau 'ppTau'
.IP "  \(bu" 4
Will not call tie while delta from null to best attrib 'ppTau', preprune if epsilon less than 'ppTau' makes sense that this be less than or equal to 'tc' (default 0, no preprune)
.PP

.IP "\(bu" 2
-initialPause
.IP "  \(bu" 4
Pause five seconds before reading the names file, sometimes needed if the file is being generated and then data is being piped to VFDT
.PP

.IP "\(bu" 2
-incrementalReporting
.IP "  \(bu" 4
As each example arrives test the learned model with it and then learn on it (default off).
.PP

.IP "\(bu" 2
-v
.IP "  \(bu" 4
Can be used multiple times to increase the debugging output.
.PP

.IP "\(bu" 2
-h
.IP "  \(bu" 4
Run vfdt -h for a list of the arguments and their meanings.
.PP

.PP

.PP
.SH SYNOPSIS
.br
.PP
.SH "Author"
.PP 
Generated automatically by Doxygen for VFML from the source code.


original/src/learners/vfdt/vfdt.c
#include "vfml.h"
#include "vfdt-engine.h"
#include <stdio.h>
#include <string.h>

#include <sys/times.h>
#include <time.h>

char *gFileStem = "DF";
char *gSourceDirectory = ".";
int   gDoTests = 0;
int   gMessageLevel = 0;

float gSplitConfidence = 0.000001;
float gTieConfidence   = 0.05;
int   gUseGini         = 0;
int   gRescans         = 1;
int   gChunk           = 300;
int   gGrowMegs        = 1000;

int   gStdin           = 0;
int   gUseSchedule     = 0;
long  gScheduleCount   = 10000;
float gScheduleMult    = 1.44;

int   gIncrementalReporting = 0;

int   gOutputTree    = 0;
int   gDoBonferonni  = 0;

int   gCacheTestSet    = 1;

int   gCacheTrainingExamples = 1;
int   gRestartScanPeriod     = 147377 * 2;
int   gRestartLeaves         = 1;

int   gDoBatch               = 0;
//int   gPrePrune              = 1;
float gPrePruneTau           = 0.00;
int   gInitialPause          = 0;

int   gLaplace               = 5;

int   gREPrune         = 0;
int   gCachePruneSet   = 0;
VoidAListPtr gPruneSet;
float        gPruneSetPercent  = .05;
int          gPruneCacheInited = 0;
long         gPruneSetMaxSize  = 1000000;
long         gPruneSetCurrentSize = 0;


static void _printUsage(char  *name) {
   printf("%s : 'Very Fast Decision Tree' induction\n", name);

   printf("-f <filestem>\t Set the name of the dataset (default DF)\n");
   printf("-source <dir>\t Set the source data directory (default '.')\n");
   printf("-u\t\t Test the learner's accuracy on data in <stem>.test\n");

   printf("-sc <prob> \t Allowed chance of error in each decision (default\n\t\t 0.0000001 that's .00001 percent)\n");
   printf("-tc <tie error>\t Call a tie when hoeffding e < this. (default\n\t\t 0.05)\n");
   printf("-gini\t\t Use gini gain instead of information gain (default\n\t\t information gain) (DOESN'T WORK FOR CONTINUOUS ATTRIBUTES)\n");
   printf("-rescans <#>\t Naievely consider each example '#' times with no\n\t\t concern for using it once per level of the induced tree\n\t\t (default 1)\n");
   printf("-chunk <count> \t Wait until 'count' examples accumulate at a\n\t\t leaf before testing for a split (default 300)\n");
   printf("-growMegs <count>\t Limit dynamic memory allocation to 'count'\n\t\t megabytes (default 1000)\n");
   printf("-stdin \t\t Reads training examples from stdin instead of from\n\t\t <stem>.data (default off) - NOTE this disables the\n\t\t rescans switch\n");
   printf("-schedule <mult> Run tests after 10k examples & every mult * 10k.\n");
   printf("-outputTree \t Output the binary tree at each test point.\n");
   printf("-noCacheTestSet \t By default vfdt reads test set from disk once\n\t\t and keeps in memory.  Use this option to conserve some memory.\n");
   printf("-prune \t Do reduced error pruning on the resulting tree\n");
   printf("-noRestartLeaves \t Do not attempt to restart leaves that were\n\t\t deactivated for memory reasons.\n");
   printf("-noCacheTrainingExamples \t Do not use extra RAM to cache training\n\t\t examples.\n");
   printf("-cachePruneExamples \t Use extra RAM to pruning examples\n\t\tonly makes sense if you use -prune (default store prune\n\t\t data in <stem>.prunedata)\n");
   printf("-batch \t Run in batch mode, read all examples into RAM, don't\n\t\t do hoeffding bounds (default off).\n");
   printf("-laplace <count> \t When starting a new node use the laplace\n\t\t correction with the parent's class and <count>\n\t\t examples (defult 5)\n");
   printf("-prePruneTau <ppTau> \t Will not call tie while delta from null\n\t\t to best attrib <ppTau>, preprune if\n\t\t epsilon < <ppTau> makes sense that this be <= tc\n\t\t (default 0, no preprune)\n");
   printf("-initialPause \t Pause five seconds before reading the names\n\t\t file, sometimes needed if the file is being generated\n\t\t and then data is being piped to VFDT\n");
   printf("-incrementalReporting \t As each example arrives test the learned\n\t\t model with it and then learn on it (default off).\n");
   printf("-v\t\t Can be used multiple times to increase the debugging output\n");
}


static void _processArgs(int argc, char *argv[]) {
   int i;

   /* HERE on the ones that use the next arg make sure it is there */
   for(i = 1 ; i < argc ; i++) {
      if(!strcmp(argv[i], "-f")) {
         gFileStem = argv[i+1];
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-source")) {
         gSourceDirectory = argv[i+1];
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-u")) {
         gDoTests = 1;
      } else if(!strcmp(argv[i], "-v")) {
         gMessageLevel++;
      } else if(!strcmp(argv[i], "-h")) {
         _printUsage(argv[0]);
         exit(0);
      } else if(!strcmp(argv[i], "-sc")) {
         sscanf(argv[i+1], "%f", &gSplitConfidence);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-tc")) {
         sscanf(argv[i+1], "%f", &gTieConfidence);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-gini")) {
         gUseGini = 1;
      } else if(!strcmp(argv[i], "-rescans")) {
         sscanf(argv[i+1], "%d", &gRescans);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-chunk")) {
         sscanf(argv[i+1], "%d", &gChunk);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-laplace")) {
         sscanf(argv[i+1], "%d", &gLaplace);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-growMegs")) {
         sscanf(argv[i+1], "%d", &gGrowMegs);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-stdin")) {
         gStdin = 1;
      } else if(!strcmp(argv[i], "-schedule")) {
         gUseSchedule = 1;
         sscanf(argv[i+1], "%f", &gScheduleMult);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-prePruneTau")) {
         sscanf(argv[i+1], "%f", &gPrePruneTau);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-outputTree")) {
         gOutputTree = 1;
      } else if(!strcmp(argv[i], "-initialPause")) {
         gInitialPause = 1;
      } else if(!strcmp(argv[i], "-bonferonni")) {
         gDoBonferonni = 1;
      } else if(!strcmp(argv[i], "-noCacheTestSet")) {
         gCacheTestSet = 0;
      } else if(!strcmp(argv[i], "-cachePruneExamples")) {
         gCachePruneSet = 1;
      } else if(!strcmp(argv[i], "-prune")) {
         gREPrune = 1;
      } else if(!strcmp(argv[i], "-noRestartLeaves")) {
         gRestartLeaves = 0;
      } else if(!strcmp(argv[i], "-noCacheTrainingExamples")) {
         gCacheTrainingExamples = 0;
      } else if(!strcmp(argv[i], "-batch")) {
	 gDoBatch = 1;
      } else if(!strcmp(argv[i], "-incrementalReporting")) {
         gIncrementalReporting = 1;
      //} else if(!strcmp(argv[i], "-noPrePrune")) {
      //   gPrePrune = 0;
      } else {
         printf("Unknown argument: %s.  use -h for help\n", argv[i]);
         exit(0);
      }
   }

   if(gMessageLevel >= 1) {
      printf("Stem: %s\n", gFileStem);
      printf("Source: %s\n", gSourceDirectory);

      printf("Split Confidence: %g\n", gSplitConfidence);
      printf("Tie Confidence: %g\n", gTieConfidence);

      if(gUseGini) {
         printf("Using Gini split index - remember this won't work for continuous attributes.\n");
      } else {
         printf("Using information gain split index\n");
      }

      if(gStdin) {
         printf("Reading examples from standard in.\n");
      } else {
         printf("Considering each example %d times.\n", gRescans);
      }
      printf("Checking for splits for every %d examples at a leaf\n", gChunk);

      if(gREPrune) {
         printf("Doing reduced error pruning on resulting tree.\n");
      }

      if(gDoTests) {
         printf("Running tests\n");
      }
   }
}

long _incrementalErrors = 0;
long _incrementalTests  = 0;
static void _doIncrementalTest(VFDTPtr vfdt, ExampleSpecPtr es, ExamplePtr e) {
   DecisionTreePtr dt;

   dt = VFDTGetLearnedTree(vfdt);
   if(ExampleGetClass(e) != DecisionTreeClassify(dt, e)) {
      _incrementalErrors++;
   }
   _incrementalTests++;

   DecisionTreeFree(dt);
}

static void _doIncrementalReport(void) {
   printf("\tTested %ld examples, made %ld mistakes that's %.4lf%%\n", 
        _incrementalTests, _incrementalErrors, 
        (float)_incrementalErrors / (float)_incrementalTests);
}

VoidAListPtr _testSet;
int          _testCacheInited = 0;
static void _doTests(ExampleSpecPtr es, DecisionTreePtr dt, long numBoundsUsed, long growingNodes, long learnCount, long learnTime, long allocation, int finalOutput) {
   int oldPool = MGetActivePool();
   char fileNames[255];
   FILE *exampleIn, *treeOut;
   ExamplePtr e;
   long i;
   long tested, errors;

   struct tms starttime;
   struct tms endtime;

   errors = tested = 0;

   /* don't track this allocation against other VFDT stuff */
   MSetActivePool(0);

   if(gREPrune) {
      if(gCachePruneSet) {
         if(!gPruneCacheInited) {
            gPruneSet = VALNew();

            sprintf(fileNames, "%s/%s.prunedata", gSourceDirectory, gFileStem);
            exampleIn = fopen(fileNames, "r");
            DebugError(exampleIn == 0, "Unable to open the .prunedata file");

            e = ExampleRead(exampleIn, es);
            while(e != 0) {
               VALAppend(gPruneSet, e);
               e = ExampleRead(exampleIn, es);
            }
            fclose(exampleIn);
            gPruneCacheInited = 1;
         }
      }

      times(&starttime);

      DebugMessage(1, 1, "Starting post pruning...\n");
      if(gCachePruneSet) {
         REPruneBatch(dt, gPruneSet);
      } else {
         sprintf(fileNames, "%s/%s.prunedata", gSourceDirectory, gFileStem);
         exampleIn = fopen(fileNames, "r");
         DebugError(exampleIn == 0, "Unable to open the .prunedata file");
         REPruneBatchFile(dt, exampleIn, 0);
         fclose(exampleIn);
      }
      DebugMessage(1, 1, "Done post pruning...\n");

      times(&endtime);
      learnTime += endtime.tms_utime - starttime.tms_utime;
      if(gMessageLevel >= 1) {
         printf("Prune Time %.2lf\n",
              ((double)endtime.tms_utime - starttime.tms_utime)/100);
      }
   }


   if(gCacheTestSet) {
      if(!_testCacheInited) {
         _testSet = VALNew();

         sprintf(fileNames, "%s/%s.test", gSourceDirectory, gFileStem);
         exampleIn = fopen(fileNames, "r");
         DebugError(exampleIn == 0, "Unable to open the .test file");

         e = ExampleRead(exampleIn, es);
         while(e != 0) {
            VALAppend(_testSet, e);
            e = ExampleRead(exampleIn, es);
         }
         fclose(exampleIn);
         _testCacheInited = 1;
      }

      for(i = 0 ; i < VALLength(_testSet) ; i++) {
         e = VALIndex(_testSet, i);
         if(!ExampleIsClassUnknown(e)) {
            tested++;
            if(ExampleGetClass(e) != DecisionTreeClassify(dt, e)) {
               errors++;
            }
         }
      }

   } else {
      sprintf(fileNames, "%s/%s.test", gSourceDirectory, gFileStem);
      exampleIn = fopen(fileNames, "r");
      DebugError(exampleIn == 0, "Unable to open the .test file");
      
      if(gMessageLevel >= 1) {
         printf("opened test file, starting scan...\n");
      }

      e = ExampleRead(exampleIn, es);
      while(e != 0) {
         if(!ExampleIsClassUnknown(e)) {
            tested++;
            if(ExampleGetClass(e) != DecisionTreeClassify(dt, e)) {
               errors++;
            }
         }
         ExampleFree(e);
         e = ExampleRead(exampleIn, es);
      }
      fclose(exampleIn);
   }

   if(finalOutput) {
      if(gMessageLevel >= 1) {
         printf("Tested %ld examples made %ld errors\n", (long)tested,
               (long)errors);
      }
      printf("%.4f\t%ld\n", ((float)errors/(float)tested) * 100,
                (long)DecisionTreeCountNodes(dt));
   } else {
      printf(">> %ld\t%.4f\t%ld\t%ld\t%.2lf\t%.2f\t%ld\n",
                learnCount,
                ((float)errors/(float)tested) * 100,
                (long)DecisionTreeCountNodes(dt),
                growingNodes,
                ((double)learnTime) / 100,
                ((double)allocation) / (1024 * 1024),
                numBoundsUsed);
   }
   fflush(stdout); 

   if(gOutputTree) {
      sprintf(fileNames, "%s-%lu.tree", gFileStem, learnCount);
      treeOut = fopen(fileNames, "w");
      DecisionTreeWrite(dt, treeOut);
      fclose(treeOut);
   }

   MSetActivePool(oldPool);
}

int main(int argc, char *argv[]) {
   char fileNames[255];

   FILE *exampleIn, *pruneSet = 0;
   ExampleSpecPtr es;
   ExamplePtr e;

   VFDTPtr vfdt;
   DecisionTreePtr dt;

   long seen = 0;

   long learnTime, allocation;

   int iteration;

   struct tms starttime;
   struct tms endtime;

   _processArgs(argc, argv);

   sprintf(fileNames, "%s/%s.names", gSourceDirectory, gFileStem);

   if(gInitialPause) {
      sleep(5);
   }

   es = ExampleSpecRead(fileNames);
   DebugError(es == 0, "Unable to open the .names file");

   RandomInit();

   /* initialize the vfdt */
   vfdt = VFDTNew(es, gSplitConfidence, gTieConfidence);
   VFDTSetUseGini(vfdt, gUseGini);
   VFDTSetProcessChunkSize(vfdt, gChunk);
   VFDTSetMaxAllocationMegs(vfdt, gGrowMegs);
   VFDTSetMessageLevel(vfdt, gMessageLevel);
   DebugSetMessageLevel(gMessageLevel);
   VFDTSetPrePruneTau(vfdt, gPrePruneTau);
   VFDTSetLaplace(vfdt, gLaplace);
   VFDTSetDoBonferonni(vfdt, gDoBonferonni);
   if(!gRestartLeaves) {
      VFDTSetRestartLeaves(vfdt, 0);
   }
   if(!gCacheTrainingExamples) {
      VFDTSetCacheTrainingExamples(vfdt, 0);
   }

   if(gMessageLevel >= 1) {
      printf("allocation %ld\n", MGetTotalAllocation());
   }

   if(gREPrune) {
      if(gCachePruneSet) {
         gPruneSet = VALNew();
         gPruneCacheInited = 1;
      } else {
         sprintf(fileNames, "%s/%s.prunedata", gSourceDirectory, gFileStem);
         pruneSet = fopen(fileNames, "w");
         DebugError(pruneSet == 0, "Unable to open the .prunedata file");
      }
   }

   if(gStdin) {
      gRescans = 1;
   }

   learnTime = 0;
   times(&starttime);

   if(gDoBatch) {
      if(gStdin) {
         exampleIn = stdin;
      } else {
         sprintf(fileNames, "%s/%s.data", gSourceDirectory, gFileStem);
         exampleIn = fopen(fileNames, "r");
         DebugError(exampleIn == 0, "Unable to open the data file");
      }

      if(gREPrune) {
         e = ExampleRead(exampleIn, es);
         while(e != 0) {
            seen++;
            if(RandomDouble() < gPruneSetPercent && 
                  gPruneSetCurrentSize < gPruneSetMaxSize) {
               gPruneSetCurrentSize++;
               if(gCachePruneSet) {
                  VALAppend(gPruneSet, e);
               } else {
                  ExampleWrite(e, pruneSet);
                  ExampleFree(e);
               }
            } else {
               VFDTProcessExampleBatch(vfdt, e);
            }
            e = ExampleRead(exampleIn, es);
         }
         VFDTBatchExamplesDone(vfdt);

      } else {
         VFDTProcessExamplesBatch(vfdt, exampleIn);
      }

      if(!gStdin) {
         fclose(exampleIn);
      }

   } else { /* not in batch mode */
      seen = 0;
      for(iteration = 0 ; iteration < gRescans ; iteration++) {
         if(gStdin) {
            exampleIn = stdin;
         } else {
            sprintf(fileNames, "%s/%s.data", gSourceDirectory, gFileStem);
            exampleIn = fopen(fileNames, "r");
            DebugError(exampleIn == 0, "Unable to open the .data file");
         }

         e = ExampleRead(exampleIn, es);

         while(e != 0) {
            seen++;

            if(gIncrementalReporting) {
               times(&endtime);
               learnTime += endtime.tms_utime - starttime.tms_utime;
   
               _doIncrementalTest(vfdt, es, e);               

               if(seen % 1000 == 0) {
                  _doIncrementalReport();
               }

               times(&starttime);
            }

            if(gREPrune && gPruneSetCurrentSize < gPruneSetMaxSize && 
                                    (RandomDouble() < gPruneSetPercent)) {
               gPruneSetCurrentSize++;
               if(gCachePruneSet) {
                  VALAppend(gPruneSet, e);
               } else {
                  ExampleWrite(e, pruneSet);
                  ExampleFree(e);
               }
            } else {
               VFDTProcessExample(vfdt, e);
            }
            /* HERE if I use an example cache I better have this commented */
            //    ExampleFree(e);

            e = ExampleRead(exampleIn, es);

            /* check to see if it's time to run tests */
            if(gUseSchedule && seen == gScheduleCount) {
               gScheduleCount *= gScheduleMult;
               allocation = MGetTotalAllocation();
               times(&endtime);
               learnTime += endtime.tms_utime - starttime.tms_utime;
               //printf("allocation %.2f\n",
                        //(float)allocation / (1024 * 1024));
               //fflush(stdout);
               dt = VFDTGetLearnedTree(vfdt);

               if(gREPrune && !gCachePruneSet) {
                  //printf("closing %d\n", fileno(pruneSet));
                  fflush(pruneSet);
                  fsync(fileno(pruneSet));
                  fclose(pruneSet);
               }
               _doTests(es, dt, VFDTGetNumBoundsUsed(vfdt), 
                      VFDTGetNumGrowing(vfdt),
                      seen, learnTime, allocation, 0);

               if(gREPrune && !gCachePruneSet) {
                  sprintf(fileNames, "%s/%s.prunedata", 
                            gSourceDirectory, gFileStem);
                  pruneSet = fopen(fileNames, "a");
                  DebugError(pruneSet == 0, "Unable to open the .prunedata file");
               }

               DecisionTreeFree(dt);
               times(&starttime);
            }
         }

         if(!gStdin) {
            fclose(exampleIn);
         }
      }
   }

   times(&endtime);
   learnTime += endtime.tms_utime - starttime.tms_utime;

   if(gMessageLevel >= 1) {
      printf("done learning...\n");

      printf("   allocation %ld\n", MGetTotalAllocation());

      printf("time %.2lfs\n", ((double)learnTime) / 100);
   }

   allocation = MGetTotalAllocation();
   dt = VFDTGetLearnedTree(vfdt);


   if(gREPrune && !gCachePruneSet) {
      //printf("closing %d\n", fileno(pruneSet));
      fflush(pruneSet);
      fsync(fileno(pruneSet));
      fclose(pruneSet);
   }


   if(gDoTests) {
      //times(&starttime);
      _doTests(es, dt, VFDTGetNumBoundsUsed(vfdt), VFDTGetNumGrowing(vfdt), seen, learnTime, allocation, 1);
   } else if(gUseSchedule) {
      //times(&starttime);
      _doTests(es, dt, VFDTGetNumBoundsUsed(vfdt), VFDTGetNumGrowing(vfdt), seen, learnTime, allocation, 0);
   } else if(gIncrementalReporting) {
      _doIncrementalReport();
   } else  {
      DecisionTreePrint(dt, stdout);
   }

   if(gMessageLevel >= 1) {
      printf("allocation %ld\n", MGetTotalAllocation());

      //times(&endtime);
      printf("time %.2lfs\n", learnTime / (float)100);

      printf("num bounds used %ld\n", vfdt->numBoundsUsed);

      printf("induced leaves by level: ");
      DecisionTreePrintStats(dt, stdout);

      printf("DONE\n");
   }

   return 0;
}

